proparse/src/main/java/org/prorefactor/core/JPNode.java

```diff
@@ -24,117 +24,465 @@
+    if (token.isNatural())
+      return this;
+    for (JPNode n = down; n != null; n = n.down) {
+      if (n.token.isNatural())
+        return n;
     }
-    for (AttributeValue attr : AttributeValue.values()) {
-      attrStrEqs.put(attr.getKey(), attr.getName());
+    return null;
+  }
+
+  /** 
+   * @return Last child of the last child of the...
+   */
+  public JPNode getLastDescendant() {
+    if (down == null)
+      return this;
+    JPNode node = down;
+    while (node.right != null) {
+      node = node.right;
     }
+    return node.getLastDescendant();
   }
 
-  public JPNode(ProToken t) {
-    this.token = t;
-    setType(t.getType());
+  /**
+   * @return First child if there is one, otherwise next sibling
+   */
+  public JPNode nextNode() {
+    return (down == null ? right : down);
   }
 
-  public String allLeadingHiddenText() {
-    String ret = "";
-    ProToken t = getHiddenFirst();
-    while (t != null) {
-      ret += t.getText();
-      t = (ProToken) t.getHiddenAfter();
+  /** 
+   * @return Previous sibling if there is one, otherwise parent
+   */
+  public JPNode getPreviousNode() {
+    return (left == null ? up : left);
+  }
+
+  // *************************
+  // End of navigation methods
+  // *************************
+
+  // ***************
+  // Various queries
+  // ***************
+
+  /** Get an ArrayList of the direct children of this node. */
+  public List<JPNode> getDirectChildren() {
+    List<JPNode> ret = new ArrayList<>();
+    JPNode n = getFirstChild();
+    while (n != null) {
+      ret.add(n);
+      n = n.getNextSibling();
     }
     return ret;
   }
 
+  /**
+   * Get an array of all descendant nodes (including this node) of a given type
+   */
+  public List<JPNode> query(ABLNodeType type, ABLNodeType... findTypes) {
+    JPNodeQuery query = new JPNodeQuery(type, findTypes);
+    walk(query);
+
+    return query.getResult();
+  }
+
+  /**
+   * Get an array of all descendant nodes (including this node) of a given type
+   */
+  public List<JPNode> queryMainFile(ABLNodeType type, ABLNodeType... findTypes) {
+    JPNodeQuery query = new JPNodeQuery(false, true, type, findTypes);
+    walk(query);
+
+    return query.getResult();
+  }
+
+  /**
+   * Get an array of all descendant nodes (including this node) of a given type
+   */
+  public List<JPNode> queryStateHead(ABLNodeType type, ABLNodeType... findTypes) {
+    JPNodeQuery query = new JPNodeQuery( true, type, findTypes);
+    walk(query);
+
+    return query.getResult();
+  }
+
+  /**
+   * Get an array of all descendant nodes (including this node) of a given type
+   */
+  @Deprecated
+  public List<JPNode> query(Integer... findTypes) {
+    JPNodeQuery query = new JPNodeQuery(findTypes);
+    walk(query);
+
+    return query.getResult();
+  }
+
+  /**
+   * Get an array of all descendant nodes (including this node) of a given type
+   */
+  @Deprecated
+  public List<JPNode> queryMainFile(Integer... findTypes) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Add the missing @deprecated Javadoc tag. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Do not forget to remove this deprecated code someday. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS1133)